### ISSUE TITLE
Include all emitted events in the contract ABI and userdoc in a non-breaking way by ignoring conflicting Natspec

### ISSUE THREAD

**Issue by cameel** [ethereum/solidity#13086]
## Abstract
Currently the contract ABI lists only events defined in the contract and inherited from other contracts. It does not list events defined in libraries that the contract emits and (after #12921) also events defined in other contracts.

Just like we do with errors, the ABI should list both defined and emitted events.

## Motivation
A contract can emit events defined in a variety of places:

```solidity
library   L { /** A library event */             event EL(uint8); }
interface I { /** An interface event */          event EI(uint16); }
contract  C { /** A foreign contract event */    event EC(uint24); }
contract  D { /** An inherited contract event */ event ED(uint32); }

contract E is D {
    /// Own event
    event EE(uint40);

    function foo() public {
        emit L.EL(1);
        emit I.EI(2);
        emit C.EC(3);
        emit ED(4);
        emit EE(5);
    }
}
```
Not all of them get listed in the ABI:
```json
[
    {
        "anonymous": false,
        "inputs":
        [{"indexed": false, "internalType": "uint32", "name": "", "type": "uint32"}],
        "name": "ED",
        "type": "event"
    },
    {
        "anonymous": false,
        "inputs":
        [{"indexed": false, "internalType": "uint40", "name": "", "type": "uint40"}],
        "name": "EE",
        "type": "event"
    },
    {
        "inputs": [],
        "name": "foo",
        "outputs": [],
        "stateMutability": "nonpayable",
        "type": "function"
    }
]
```

This poses a problem for tools like Truffle, which use the ABI to determine which events the contract might emit.

The `solc --userdoc` output compiled from the Natspec docstrings also only lists this subset of events:
```json
{
    "events":
    {
        "ED(uint32)": {"notice": "An inherited contract event "},
        "EE(uint40)": {"notice": "Own event"}
    },
    "kind": "user",
    "methods": {},
    "version": 1
}
```

## Specification
The only real problem here is the userdoc. The current structure only allows one set of docs for each signature. We are planning to change the structure in Natspec v2 (#11114) but this will be a breaking change.

Since not including these events in the ABI is so problematic for tools, we have decided to simply ignore the Natspec problem - for now in case of conflicting signatures we'll include the Natspec docstring only if there is one in the current contract (defined directly in it or inherited). Otherwise we'll ignore the docstring. This is not the correct way to handle this but we expect the impact of this problem to be very low and the problem will disappear once we release Natspec v2 anyway.

## Backwards Compatibility
Unlike #11114, this is fully backwards compatible in terms of ABI and userdoc structure. It does change the meaning of event list included there but not in a way that should pose problems to tools.

**Comment by cameel** [comment]
One more thing that we discussed on the call today is that this conflict that results in one of the events being ignored should trigger a warning. This was with the assumption that the conflict happens both for ABI and for Natspec. Given that the ABI is fine after all, I think the warning is not necessary.

**Comment by cameel** [comment]
Also, a note to anyone who will be implementing this: we already have a PR for this (#10996) and it's mostly done. It will just need to be split into the part that can go into `develop` without compatibility issues and one that must go to `breaking`.

**Comment by cameel** [comment]
We decided to postpone the change I mentioned above until 0.8.16. For 0.8.15 we'll instead revert the problematic part of #12921. Here's my PR that does this: #13135.

For 0.8.16 we should revert the first commit from that PR.

**Comment by PaulRBerg** [comment]
Sorry to chase, but I'm wondering if there is an ETA for when this issue will be fixed?

Our current set-up looks like this:

```tree
├── libraries
│  ├── Errors.sol
│  └── Events.ol
├── Contract1.sol
└── Contract2.sol
```

Even if the contract reverts with the errors and emit the events defined in the libraries, the contract ABIs don't contain any errors and/ or events, so our front-end engineers have to manually copy-and-paste the error and the events ABIs.

**Comment by frangio** [comment]
Just ran into this as well.

> in case of conflicting signatures we'll include the Natspec docstring only if there is one in the current contract

This sounds good to me.

**Comment by cameel** [comment]
> Sorry to chase, but I'm wondering if there is an ETA for when this issue will be fixed?

We have a PR in review currently and I think we managed to whittle our PR backlog down enough to get back to it soon. Won't make it into the very next release because that's too soon but 0.8.19 is likely.

**Comment by PaulRBerg** [comment]
Awesome, thanks a lot, @cameel!

**Comment by PaulRBerg** [comment]
Does anyone know if this issue has any implications for source code verification on Etherscan? I worry that there might be a mismatch between the actual ABI and the one reported by Etherscan.

Presumably, Etherscan would use the solc compiler output as is, and not polyfill the missing events and errors on behalf of the users.

**Comment by cameel** [comment]
No idea. Are they known to actually add stuff to the ABI they publish on their site?

**Comment by PaulRBerg** [comment]
[It](url) looks like this is a major issue with Etherscan. I just deployed this contract to Goerli (using Forge):

```solidity
pragma solidity >=0.8.18;

library Events {
    event LogAddress(address addr);
}

contract Foo {
    function logAddress(address addr) external {
        emit Events.LogAddress(addr);
    }
}
```

I then proceeded to manually verify it via the Etherscan UI, where I got this error (I made sure that I have the same compiler settings):

> Error! Unable to generate Contract ByteCode and ABI

You can see the full error [here](https://imgur.com/a/iPKQgWS).

Verifying via Forge worked:

> forge verify-contract --chain 5 --optimizer-runs 10000 0x4719e95a2fAc85Cddd7cf2Bc2De4D229E1Ec044F src/Foo.sol:Foo

Now, here's the issue. The ABI on the [Etherscan contract page](https://goerli.etherscan.io/address/0x4719e95a2fac85cddd7cf2bc2de4d229e1ec044f#code) does NOT include the `LogAddress` event. It only lists out the `logAddress` function:

```json
[{"inputs":[{"internalType":"address","name":"addr","type":"address"}],"name":"logAddress","outputs":[],"stateMutability":"nonpayable","type":"function"}]
```

Therefore, until this bug is fixed in the compiler itself, all Solidity contracts verified on Etherscan will lack the events that the said contract emits, which will cause coordinations issues between contract developers and third-party integrators. It is unlikely that Etherscan would be interested in building an ad-hoc solution for adding all the missing events in the ABI - building such a solution would take time, and it would be hard to account for all possible edge cases.

I think that the priority for this issue should be set to high - it would be great to have it fixed in in v0.8.19.

**Comment by PaulRBerg** [comment]
It looks like Solidity v0.8.19 dropped yesterday but did not include a fix for this bug. I can understand the rationale, though - you wanted to get the user-defined operators out there, and fix the more serious dead code bug.

Will a fix for this bug be included in Solidity v0.8.20?

I'm sorry for being a broken record in this GitHub issue. It's just that this bug prevents us from going to mainnet. We won't be able to deploy until this is fixed because we don't want our verified contracts to lack the events defined in our libraries.

**Comment by cameel** [comment]
> Will a fix for this bug be included in Solidity v0.8.20?

I can't promise anything, but it's very possible that we'll be done with this before the next release. It's just that we're still not fully done with UDVTs - we still need to implement literal suffixes and we can't afford to have it slip beyond Q1. If we're squeezed for time, unfortunately this issue might end up taking the hit again.

Now, if it really is the cause of verification problems we should definitely bump the priority here. It's just that it has been working like this for ages so it's a bit weird that a problem like this would suddenly come up now. This is why it's been considered just medium impact for now - quite important but not the end of the world if it has to be delayed.

I don't really understand the reason behind the Etherscan failure though. How does not including an event in the ABI cause the verification to fail? If both the tool and Etherscan are using the same version of the compiler and the same sources, both should be getting the same bytecode and ABI.

**Comment by PaulRBerg** [comment]
> but it's very possible that we'll be done with this before the next release

This would be amazing 🙏

> How does not including an event in the ABI cause the verification to fail?

Sorry for not being clear enough before.

The issue is not that the verification fails, but that verification passes in a _misleading_ way. That is, the official ABIs that would appear on our Etherscan contract profiles (and any other blockchain explorers, for that matter) would NOT contain any events, because we source all of our events from a shared library called `Events.sol`.

> both should be getting the same bytecode and ABI.

Yep, this is what happens. But both would see an ABI that is missing events.

**Comment by PaulRBerg** [comment]
I want to include @Amxx in this discussion, since he might be able to shed more light on the importance of fixing this bug.

My understanding is that because of this bug, they (OpenZeppelin) [have resorted to using abstract contracts instead of libraries](https://twitter.com/Amxx/status/1622206396778283008) in their V5 library (which is still being worked on at the moment).

**Comment by Amxx** [comment]
> I want to include @Amxx in this discussion, since he might be able to shed more light on the importance of fixing this bug.
> 
> My understanding is that because of this bug, they (OpenZeppelin) [have resorted to using abstract contracts instead of libraries](https://twitter.com/Amxx/status/1622206396778283008) in their V5 library (which is still being worked on at the moment).

That is already the case in our 4.x contracts!

This should be a library https://github.com/OpenZeppelin/openzeppelin-contracts/blob/master/contracts/proxy/ERC1967/ERC1967Upgrade.sol
